timesketch/lib/utils.py

```diff
@@ -101,79 +117,59 @@ def read_and_validate_csv(file_handle, delimiter=','):
     if not isinstance(delimiter, six.text_type):
         delimiter = codecs.decode(delimiter, 'utf8')
 
-    # Due to issues with python2.
-    if six.PY2:
-        delimiter = str(delimiter)
-
-    reader = csv.DictReader(file_handle, delimiter=delimiter)
-    csv_header = reader.fieldnames
-    missing_fields = []
-    # Validate the CSV header
-    for field in mandatory_fields:
-        if field not in csv_header:
-            missing_fields.append(field)
-    if missing_fields:
-        raise RuntimeError(
-            'Missing fields in CSV header: {0:s}'.format(
-                ','.join(missing_fields)))
     try:
-        for row in reader:
-            # There is a condition in which the CSV reader can read a
-            # single lines as multiple, causing issues with importing.
-            # TODO: Swap the CSV library for the user of pandas.
-            if not row:
-                continue
-            if not row['datetime']:
-                logger.warning(
-                    'Row missing a datetime object, skipping [{0:s}]'.format(
-                        ','.join([str(x).replace(
-                            '\n', '').strip() for x in row.values()])))
-                continue
-            try:
-                # normalize datetime to ISO 8601 format if it's not the case.
-                parsed_datetime = parser.parse(row['datetime'])
-                row['datetime'] = parsed_datetime.isoformat()
-
-                normalized_timestamp = int(
-                    time.mktime(parsed_datetime.utctimetuple()) * 1000000)
-                normalized_timestamp += parsed_datetime.microsecond
-                row['timestamp'] = str(normalized_timestamp)
-                if 'tag' in row:
-                    row['tag'] = [x for x in _parse_tag_field(row['tag']) if x]
-
-                _scrub_special_tags(row)
-            except ValueError:
-                continue
-
-            yield row
-    except csv.Error as e:
+        reader = pandas.read_csv(file_handle, sep=delimiter)
+    except pandas.errors.ParserError as e:
         error_string = 'Unable to read file, with error: {0!s}'.format(e)
         logger.error(error_string)
         raise errors.DataIngestionError(error_string)
 
+    validate_csv_fields(mandatory_fields, reader)
+    for row in reader:
+        if not row:
+            continue
+        if not row['datetime']:
```

Comment:
I would opt not to do this in the iterator... 

rather do the whole checks for datetime and other fields with the entire data frame, here you can do the checks and parsing on the entire frame instead of a single row.